lib/realtime_web/channels/realtime_channel.ex

```diff
@@ -6,213 +6,97 @@ defmodule RealtimeWeb.RealtimeChannel do
 
   require Logger
 
+  alias Realtime.Helpers
+
   alias DBConnection.Backoff
+
   alias Phoenix.Tracker.Shard
-  alias RealtimeWeb.{ChannelsAuthorization, Endpoint, Presence}
-  alias Realtime.{GenCounter, RateCounter, PostgresCdc, SignalHandler, Tenants}
-
-  import Realtime.Helpers, only: [cancel_timer: 1, decrypt!: 2]
-
-  defmodule Assigns do
-    @moduledoc false
-    defstruct [
-      :tenant,
-      :log_level,
-      :rate_counter,
-      :limits,
-      :tenant_topic,
-      :pg_sub_ref,
-      :pg_change_params,
-      :postgres_extension,
-      :claims,
-      :jwt_secret,
-      :tenant_token,
-      :access_token,
-      :postgres_cdc_module,
-      :channel_name
-    ]
 
-    @type t :: %__MODULE__{
-            tenant: String.t(),
-            log_level: atom(),
-            rate_counter: RateCounter.t(),
-            limits: %{
-              max_events_per_second: integer(),
-              max_concurrent_users: integer(),
-              max_bytes_per_second: integer(),
-              max_channels_per_client: integer(),
-              max_joins_per_second: integer()
-            },
-            tenant_topic: String.t(),
-            pg_sub_ref: reference() | nil,
-            pg_change_params: map(),
-            postgres_extension: map(),
-            claims: map(),
-            jwt_secret: String.t(),
-            tenant_token: String.t(),
-            access_token: String.t(),
-            channel_name: String.t()
-          }
-  end
+  alias Realtime.GenCounter
+  alias Realtime.PostgresCdc
+  alias Realtime.RateCounter
+  alias Realtime.SignalHandler
+  alias Realtime.Tenants
+  alias Realtime.Tenants.Connect
+
+  alias RealtimeWeb.ChannelsAuthorization
+  alias RealtimeWeb.Endpoint
+  alias RealtimeWeb.Presence
 
   @confirm_token_ms_interval 1_000 * 60 * 5
 
   @impl true
-  def join(
-        "realtime:" <> sub_topic = topic,
-        params,
-        %{
-          assigns: %{
-            tenant: tenant,
-            log_level: log_level,
-            postgres_cdc_module: module
-          },
-          channel_pid: channel_pid,
-          serializer: serializer,
-          transport_pid: transport_pid
-        } = socket
-      ) do
+  def join("realtime:" <> sub_topic = topic, params, socket) do
+    %{
+      assigns: %{tenant: tenant, log_level: log_level, postgres_cdc_module: module},
+      channel_pid: channel_pid,
+      serializer: serializer,
+      transport_pid: transport_pid
+    } = socket
+
     Logger.metadata(external_id: tenant, project: tenant)
     Logger.put_process_level(self(), log_level)
 
-    socket = socket |> assign_access_token(params) |> assign_counter()
+    socket =
+      socket
+      |> assign_access_token(params)
+      |> assign_counter()
 
     start_db_rate_counter(tenant)
 
     with false <- SignalHandler.shutdown_in_progress?(),
+         {:ok, _} <- Connect.connection_status(tenant),
```

Comment:
Am I correct in understanding that db availability is checked only once when `:syn.lookup(__MODULE__, tenant_id)` returns undefined?

Reply:
yes, we will check once that we are able to connect. this is actually a problem since I'm setting syn also on connect error and that shouldn't happen because that way if one user, on join, is unable to connect a second user will retry it and check again 👀 

this in a way could be the retry policy "retry if syn is undefined" and "success if syn is set". this will simplify the full logic and was something Chasers mentioned so I will go with that 😄